keps/sig-storage/2639-secret-protection/README.md

```diff
@@ -0,0 +1,531 @@
+
+#### Story 3
+
+A user really would like to delete a secret despite that it is used by other resources.
+The user force delete the secret while it is used by other resources, and the secret isn't protected and is actually deleted.
+
+### Notes/Constraints/Caveats (Optional)
+
+- Compatibility:
+  - There might be many existing scripts that don't care the order of deletion. Therefore, such scripts might stuck on Secret/ConfigMap deletion, if the deletion of the resources using the Secrets/ConfigMaps are done later.
+- Usability:
+  - Use of the Secret/ConfigMap in other resource will not be obvious to users. Therefore, users might not easily understand why the Secret/ConfigMap is not deleted.
+  - Users might need to force delete the Secret/ConfigMap on deletion or would like to avoid protection for certain Secrets/ConfigMaps that already exist or that are newly created.
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+There is a corner case when Reclaim Policy is Retain and PV is referencing secrets in a certain user namespace.
+In such a case, when PVC is deleted, the PV referencing secrets remains as a resource not managed by the namespace, as a result, deletion of the secrets is blocked.
+However, the user of the namespace won't notice why it is blocked and won't find a way to delete the secrets other than manually deleting finalizer.
+In addition, to make things worse, provisioner secrets that are defined by using template with PVC information,
+like "${pvc.name}", would be deleted in above case, due to the failure in finding a reference to the secret,
+because the reference to provisioner secret is not stored in PV and needs to be resolved on deletion time with PVC information that is already deleted.
+This behavior is inconsistent with other secrets, like node publish secret whose reference is stored in PV.
+
+## Design Details
+
+This feature can be implemented in the same way as `pv-protection/pvc-protection`.
+Protection logics for in-tree resources and out-of-tree resources are separeted and independently work.
+It is due to the restriction that CRDs can't be handled from in-tree controller. Note that in-tree controller protect both Secrets and ConfigMaps, on the other hand, out-of-tree controller protect only Secrets because VoluemSnapshots never use ConfigMaps.
+
+- In-tree resources(`Pod` and `PersistentVolume`):
+  - The deletion is blocked by using newly introduced `kubernetes.io/ephemeral-protection` finalizer,
+  - The `kubernetes.io/ephemeral-protection` finalizer will be always added on creation of the Secret/ConfigMap by using admission controller for in-tree resources,
```

Comment:
[This](https://github.com/kubernetes/kubernetes/blob/master/plugin/pkg/admission/storage/storageobjectinuseprotection/admission.go#L67) is how current pv-protection and pvc-protection works. The finalizer [document]( https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#finalizers) doesn't seem to ask adding finalizer to only resources that are actually used. Is it just a design choice that you recommend or are there any reasons to avoid it?

Reply:
> This is how current pv-protection and pvc-protection works. The finalizer document doesn't seem to ask adding finalizer to only resources that are actually used.

PV and PVC resources are only used for volume storage, so expecting them to be managed by a storage finalizer is reasonable.

Secrets and configmaps are used broadly for many different purposes. Adding finalizers proactively to secrets that are not yet used increases the risk of deadlocks and disrupting existing API clients managing those resources.

Reply:
If it really matters, I would make this feature an opt-in style.
I mean making this feature work only when `secret.kubernetes.io/enable-secret-protection: "yes"` annotation is added or adding a new type of [Secret](https://kubernetes.io/docs/concepts/configuration/secret/#secret-types) for this purpose.

I thought that guarantee of the right order of deleting resources should be a basic feature that is provided as a default, so I made it as an opt-out style from a UX viewpoint. 

In addition, from an implementation view point, I'm worrying about failure in adding the finalizer in corner cases by relying on the state to decide whether to add the finalizer. For example, when secret and consuming resource is created almost at the same time, can we guarantee that the finalizer is added to the secret? I think that finalizer is just a way to add a pre-delete hook, so it should be OK that controller just handles it properly on deletion time and avoid complex logics on adding finalizers.

WDYT?

Reply:
@liggitt 

According to the [discussion](https://github.com/kubernetes/enhancements/pull/2840#pullrequestreview-716197279), it looks like that without lien, it is difficult to avoid pre-emptively adding finalizers with providing a guarantee of the protection. While I understand that we should minimize risks that issues happen to unrelated resources by unnecessary finalizers.

Therefore, I would like to continue working on implementing secret protection with opt-in style, explained above. It would also minimize the risks.

I will update the KEP and ask you for review.

Reply:
I do not agree with preemptively adding finalizers for this use case. Add them when the use is noticed. In this case there's a race and that's fine. (However, I don't agree with the premise of this KEP, I will leave another comment.)